### ISSUE TITLE
Allow aggregate initialisation of components?

### ISSUE THREAD

**Issue by radman0x** [alecthomas/entityx#212]
https://github.com/alecthomas/entityx/blob/6389b1f91598c99d85e56356fb57d9f4683071d8/entityx/Entity.h#L648

Seems like the line I've tagged could be easily changed to use brace initialisation rather than direct initialisation. I'm wondering if this has been considered previously and any reason for the choice to avoid it.

Brace initialisation would allow for [aggregate initialisation](https://en.cppreference.com/w/cpp/language/aggregate_initialization) meaning that you don't need to manually create a constructor for each component. Given that components should frequently qualify as aggregates this would seem to be a nice improvement by removing the need to create constructors unnecessarily.

Thoughts?

**Comment by alecthomas** [comment]
Yes this has been considered before, but I believe it would break API compatibility.

**Comment by alecthomas** [comment]
See #87.

**Comment by alecthomas** [comment]
I wonder if it's possible to achieve backwards compatibility with SFINAE...

**Comment by radman0x** [comment]
Hmmm I suspect trying to SFINAE to the correct types would be a direct clash with the use of variadic arguments, but that's just my initial feel.

In general though the narrowing conversion issues are pretty easy to get around. They only really occur when using literals or ambiguous temporaries. This can be resolved by ensuring the right types are passed in e.g. int(5) rather than just 5 or std::string("oeau") instead of "oeau". Still would be a major compatibility issue for existing code bases. For me though I've made the change locally as removing the need to write endless redundant constructors comes with almost no cost. I think the change is a good one in general for any new project using EntityX.

**Comment by alecthomas** [comment]
Thanks @FrankStain .

You know what, let me tag a stable version, then if you feel like sending a PR with your proposed change @radman0x, I'll merge it. Better to keep adapting it then stay static and annoying.

**Comment by alecthomas** [comment]
Okay, tagged 1.3.0.

**Comment by radman0x** [comment]
@alecthomas 
I'll look at sending my raw changes through in a PR, I've also found a second location where brace initialisation needed to be added. Changes are trivial, literally just changing the parens to braces.

@FrankStain 
Thanks for pointing out `is_aggregate` and `is_constructible` but I'm not sure that it will be able to remove the narrowing conversion issues. Brace initialisation is uniform and can be used with aggregates or to call defined constructors transparently. I can't see what adding in traits to specialise separately can add as the issue is the types of the parameters being passed to the constructor (or aggregate init) rather than the syntax used to trigger construction. So the issue is getting the types of the Construct arguments to match exactly what the aggregate expects... Maybe I've missed something, feel free to elaborate if that's the case, I'm still mulling over if there's any other way.

As an aside I think usage of `is_aggregate` is out of reach for the moment as I don't think VS supports it and from what I can gather g++ needs 7.1 or above :(

**Comment by radman0x** [comment]
Ahhh I see, haha it's always right after you finish a response :P Specialising for aggregates would ease the backwards compatibility issues as any current code won't use brace initialisation as this is reserved only for aggregates.

**Comment by alecthomas** [comment]
Don't worry about backwards compatibility.

**Comment by kim366** [comment]
Why was it reverted? Couldn’t find any explanation
